src/TestPlan.lua

```diff
@@ -2,68 +2,259 @@
+
+	env.expect = Expectation.new
+
+	return env
 end
 
+local TestNode = {}
+TestNode.__index = TestNode
+
 --[[
-	Creates a new node that would be suitable to insert into the TestPlan.
+	Create a new test node. A pointer to the test plan, a phrase to describe it
+	and the type of node it is are required. The modifier is optional and will
+	be None if left blank.
 ]]
-function TestPlan.createNode(phrase, nodeType, nodeModifier)
+function TestNode.new(plan, phrase, nodeType, nodeModifier)
 	nodeModifier = nodeModifier or TestEnum.NodeModifier.None
 
 	local node = {
+		plan = plan,
 		phrase = phrase,
 		type = nodeType,
 		modifier = nodeModifier,
 		children = {},
 		callback = nil,
-		getFullName = constructNodeFullName
+		parent = nil,
+	}
+
+	node.environment = newEnvironment(node, plan.extraEnvironment)
+	return setmetatable(node, TestNode)
+end
+
+function TestNode:addChild(phrase, nodeType, nodeModifier)
+	if self.plan.testNamePattern and (nodeModifier == nil or nodeModifier == TestEnum.NodeModifier.None) then
+		local name = self:getFullName() .. " " .. phrase
+		if name:match(self.plan.testNamePattern) then
+			nodeModifier = TestEnum.NodeModifier.Focus
+		else
+			nodeModifier = TestEnum.NodeModifier.Skip
+		end
+	end
+	local child = TestNode.new(self.plan, phrase, nodeType, nodeModifier)
+	child.parent = self
+	table.insert(self.children, child)
+	return child
+end
+
+--[[
+	Join the names of all the nodes back to the parent.
+]]
+function TestNode:getFullName()
+	if self.parent then
+		local parentPhrase = self.parent:getFullName()
+		if parentPhrase then
+			return parentPhrase .. " " .. self.phrase
+		end
+	end
+	return self.phrase
+end
+
+--[[
+	Expand a node by setting its callback environment and then calling it. Any
+	further it and describe calls within the callback will be added to the tree.
+]]
+function TestNode:expand()
+	local originalEnv = getfenv(self.callback)
+	local callbackEnv = setmetatable({}, { __index = originalEnv })
+	for key, value in pairs(self.environment) do
+		callbackEnv[key] = value
+	end
+	setfenv(self.callback, callbackEnv)
+
+	local success, result = xpcall(self.callback, function(err)
+		return err .. "\n" .. debug.traceback()
+	end)
+
+	if not success then
+		self.loadError = result
+	end
+end
+
+local TestPlan = {}
+TestPlan.__index = TestPlan
+
+--[[
+	Create a new, empty TestPlan.
+]]
+function TestPlan.new(testNamePattern, extraEnvironment)
+	local plan = {
+		children = {},
+		testNamePattern = testNamePattern,
+		extraEnvironment = extraEnvironment,
 	}
 
-	return node
+	return setmetatable(plan, TestPlan)
+end
+
+--[[
+	Add a new child under the test plan's root node.
+]]
+function TestPlan:addChild(phrase, nodeType, nodeModifier)
+	if self.testNamePattern and (nodeModifier == nil or nodeModifier == TestEnum.NodeModifier.None) then
+		if phrase:match(self.testNamePattern) then
+			nodeModifier = TestEnum.NodeModifier.Focus
+		else
+			nodeModifier = TestEnum.NodeModifier.Skip
+		end
+	end
```

Comment:
This logic is duplicated between here and `TestNode`. Should we factor it out?

Reply:
Will do

Reply:
Done